cmd/runtimetest/main.go

```diff
@@ -370,6 +371,52 @@ func validateMountsExist(spec *rspec.Spec) error {
 	return nil
 }
 
+func getActualCgroupPath() (string, error) {
+	f, err := os.Open("/proc/self/cgroup")
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+
+	s := bufio.NewScanner(f)
+	for s.Scan() {
+		if err := s.Err(); err != nil {
+			return "", err
+		}
+
+		text := s.Text()
+		parts := strings.Split(text, ":")
+		return parts[2], nil
```

Comment:
On Tue, Aug 02, 2016 at 02:18:35AM -0700, Ma Shimiao wrote:
> @wking The Spec says `If a cgroupsPath value is specified, the
> runtime MUST consistently attach to the same place in the cgroup
> hierarchy given the same value of cgroupsPath.`

This is “for multiple runtime invocations”, which is something you can
test via a test harness, but not something you can test by
runtimetest, which only knows about the container it is inside.

> So, I think all controllers'path in /proc/self/cgroup must be
> same. If not, the validation should be fail. If it is, return the
> path. There is no need to be so complicated.

The reason controllers touched by the runtime need to use the same
path is [1]:

  The Spec does not support per-controller paths for the reasons
  discussed in the [cgroupv2 documentation][cgroup-v2].

The reason the host paths are also an option is [2]:

  A runtime MUST at least use the minimum set of cgroup controllers
  required to fulfill the `resources` settings.  However, a runtime
  MAY attach the container process to additional cgroup controllers
  supported by the system.

which means the runtime is free to ignore controllers that aren't
related to the configured ‘linux.resources’.  Those ignored
controllers will inherit whatever paths the host process had.

[1]: https://github.com/opencontainers/runtime-spec/pull/493/files#diff-428eec4013a655816cdefafd5d3505f1R171
[2]: https://github.com/opencontainers/runtime-spec/pull/493/files#diff-428eec4013a655816cdefafd5d3505f1R178